source/extensions/filters/http/jwt_authn/verifier.h

```diff
@@ -0,0 +1,72 @@
+#pragma once
+
+#include "extensions/filters/http/jwt_authn/authenticator.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace HttpFilters {
+namespace JwtAuthn {
+
+class Verifier;
+typedef std::unique_ptr<Verifier> VerifierPtr;
+
+typedef std::function<AuthenticatorPtr(const std::vector<std::string>& audiences,
+                                       const absl::optional<std::string>& issuer,
+                                       std::vector<JwtLocationConstPtr>&& tokens)>
+    AuthFactory;
+
+class VerifierContext;
+typedef std::unique_ptr<VerifierContext> VerifierContextPtr;
+
+/**
+ * Supports verification of JWTs with configured requirments.
+ */
+class Verifier {
+public:
+  virtual ~Verifier() {}
+  /**
+   * Object used to signal the caller of completion.
+   */
+  class Callbacks {
+  public:
+    virtual ~Callbacks() {}
+
+    virtual void onComplete(const ::google::jwt_verify::Status& status) PURE;
+  };
+  // Verify all tokens on headers, and signal the caller with callback.
+  virtual void verify(VerifierContext& context) PURE;
+
+  // Factory method for creating verifiers.
+  static VerifierPtr
+  create(const ::envoy::config::filter::http::jwt_authn::v2alpha::JwtRequirement& requirement,
+         const Protobuf::Map<ProtobufTypes::String,
+                             ::envoy::config::filter::http::jwt_authn::v2alpha::JwtProvider>&
+             providers,
+         const AuthFactory& factory);
+
+  static VerifierContextPtr createContext(Http::HeaderMap& headers, Callbacks* callback);
+};
+
+/**
+ * Contains authenticator and callbacks related data which are created per request.
+ */
+class VerifierContext {
+public:
+  virtual ~VerifierContext() {}
+
+  // Complete this verifier context.
+  virtual void complete(const ::google::jwt_verify::Status& status) PURE;
+  // Return if this context is completed.
+  virtual bool isCompleted() const PURE;
+  // add inner context.
+  virtual void add(VerifierContextPtr&& inner_context) PURE;
```

Comment:
This is very odd!!!  I don't like it

Reply:
it is mirroring verifier hierarchy, and setting up levels of callbacks.